massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -369,11 +369,15 @@ impl SpeculativeLedger {
     /// # Returns
     /// `Some(Vec<Vec<u8>>)` for found keys, `None` if the address does not exist.
     pub fn get_keys(&self, addr: &Address) -> Option<BTreeSet<Vec<u8>>> {
-        // here, get the final keys from the final ledger
-        let mut keys: Option<BTreeSet<Vec<u8>>> = self.final_state
-            .read()
-            .ledger
-            .get_datastore_keys(addr);
+
+        let mut keys: Option<BTreeSet<Vec<u8>>> = match self.final_state.read().ledger.entry_exists(addr) {
+            // here, get the final keys from the final ledger
+            true => self.final_state
```

Comment:
problem: this is going to deadlock (final_state is locked in the scope above)

Reply:
Does it apply also for parking_lot::RwLock (we are using this one and not the std::sync::RwLock impl)?

Reply:
Code updated anyway!

Reply:
> Does it apply also for parking_lot::RwLock (we are using this one and not the std::sync::RwLock impl)?

In parking_lot, it does not panic, but deadlocks: `Note that attempts to recursively acquire a read lock on a RwLock when the current thread already holds one may result in a deadlock.`

In any case, it is good practice to minimize locks, and avoid recursive locks in the same thread. When it is absolutely necessary,  re-entrant locks should be used: https://docs.rs/parking_lot/latest/parking_lot/type.ReentrantMutex.html